File: Homework/Chau_Long/lesson_2/sort_list.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head

        fast, slow = head.next, head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next

        head_right = slow.next
        slow.next = None
        left, right = self.sortList(head), self.sortList(head_right)
        return self.merge(left, right)

    def merge(self, h1: Optional[ListNode], h2: Optional[ListNode]) -> Optional[ListNode]:
        if not h1 or not h2:
            return h1 or h2

        dummy = cur = ListNode()
        while h1 and h2:
            if h1.val > h2.val:
                cur.next = h2
                h2 = h2.next
            else:
                cur.next = h1
                h1 = h1.next
            cur = cur.next
        cur.next = h1 or h2

        return dummy.next

File: Homework/Duy_Nguyen/day_3/SortList.java
package org.global.dev.day_3;

import org.global.dev.common.ListNode;

public class SortList {
	
	
	public ListNode sortList(ListNode head) {
		if (head == null || head.next == null) {
			return head;
		}
		ListNode mid = getMid(head);
		ListNode left = sortList(head);
		ListNode right = sortList(mid);
		return merge(left, right);
	}
	
	// 1 -> 5 -> 7
	// 2 -> 3 -> 8
	
	public ListNode merge(ListNode list1, ListNode list2) {
		ListNode head = new ListNode();
		ListNode run = head;
		
		while (list1 != null && list2 != null) {
			
			if (list1.val < list2.val) {
				run.next = list1;
				list1 = list1.next;
			} else {
				run.next = list2;
				list2 = list2.next;
			}
			run = run.next;
		}
		
		run.next = (list1 != null) ? list1 : list2;
		return head.next;
	}
	
	public ListNode getMid(ListNode head) {
		ListNode slow = null;
		ListNode fast = head;
		
		while (fast != null && fast.next != null) {
			slow = (slow == null) ? head : slow.next;
			fast = fast.next.next;
		}
		
		assert slow != null;
		ListNode mid = slow.next;
		slow.next = null;
		return mid;
	}
	
}
